src/coreclr/src/vm/diagnosticsipcfactory.cpp

```diff
@@ -0,0 +1,152 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "diagnosticsprotocol.h"
+#include "diagnosticsipcfactory.h"
+
+#ifdef FEATURE_PERFTRACING
+
+IpcStream **DiagnosticsIpcFactory::s_ppActiveConnectionsCache = nullptr;
+uint32_t DiagnosticsIpcFactory::s_ActiveConnectionsCacheSize = 0;
+
+IpcStream::DiagnosticsIpc *DiagnosticsIpcFactory::CreateServer(const char *const pIpcName, ErrorCallback callback)
+{
+    return IpcStream::DiagnosticsIpc::Create(pIpcName, IpcStream::DiagnosticsIpc::ConnectionMode::SERVER, callback);
+}
+
+IpcStream::DiagnosticsIpc *DiagnosticsIpcFactory::CreateClient(const char *const pIpcName, ErrorCallback callback)
+{
+    return IpcStream::DiagnosticsIpc::Create(pIpcName, IpcStream::DiagnosticsIpc::ConnectionMode::CLIENT, callback);
+}
+
+IpcStream *DiagnosticsIpcFactory::GetNextAvailableStream(IpcStream::DiagnosticsIpc *const *const ppIpcs, uint32_t nIpcs, ErrorCallback callback)
+{
+    // a static array that holds open client connections that haven't been used
+    // Remove entries from this list that have been used, e.g., they are placed in pStream and returned
+    // This will prevent the runtime from continually reestablishing connection when the server loop loops.
+    // This does, however, introduce state to this method which is undesireable, but a justifiable cost to minimizing system calls.
+
+    if (s_ppActiveConnectionsCache == nullptr)
+    {
+        ResizeCache(nIpcs);
+    }
+
+    if (s_ActiveConnectionsCacheSize != nIpcs)
+    {
+        // number of connections has changed
+        // (3/2020 - This isn't possible, but should be here for future proofing)
+        ClearCache();
+        ResizeCache(nIpcs);
+    }
+
+    IpcStream *pStream = nullptr;
+    
+    // Polling timeout semantics
+    // If client connection is opted in
+    //   and connection succeeds => set timeout to max
+    //   and connection fails => set timeout to minimum and scale by falloff factor
+    // else => set timeout to -1 (infinite)
+    //
+    // If an agent closes its socket while we're still connected,
```

Comment:
Can't we detect the disconnect via WaitForMultipleObjects()/poll()? 

More generally we should be striving that an idle .Net app isn't polling anything. I'd be OK polling if the agent pipe is missing and the environment variable told us we should expect its presence but I am hoping that is the only case it happens.

Reply:
Changed it so on successful connection we don't poll.  `IpcStreamFactory::Poll` will return if a connection closes remotely, e.g., Agent goes down etc.  Polling will only happen when a connection is supposed to be there and it isn't.